libraries/AP_PacketDigitalCAN/AP_PacketDigitalCAN.cpp

```diff
@@ -0,0 +1,609 @@
+            send_command(PacketType::ALGORITHM_SET, MPPT_devices[i].serialnumber, _config.algorithm);
+        }
+        break;
+    }
+}
+
+// parse inbound frames
+void AP_PacketDigitalCAN::handle_frame(const AP_HAL::CANFrame &frame)
+{
+    const uint16_t serialnumber = frame.id & 0x0000FFFF;
+    if (serialnumber == 0) {
+        // This is for broadcast and I don't think we should allow this inbound.
+        return;
+    }
+
+    WITH_SEMAPHORE(_sem);
+
+    uint8_t index = get_device_index(serialnumber);
+    if (index == UINT8_MAX) {
+        // we don't know this device
+        if (device_count >= device_count_max) {
+            // we don't have any room to remember it
+            return;
+        }
+        // add it
+        index = device_count;
+        device_count++;
+        MPPT_devices[index].serialnumber = serialnumber;
+        //gcs().send_text(MAV_SEVERITY_DEBUG, "PDCAN: %u New device", (unsigned)serialnumber);
+
+        send_command(PacketType::VOLTAGE_GET, serialnumber);
+        send_command(PacketType::ALGORITHM_GET, serialnumber);
+        send_command(PacketType::CVT_GET, serialnumber);
+    }
+
+    const uint8_t sequence = frame.data[0];
+    //gcs().send_text(MAV_SEVERITY_DEBUG, "PDCAN: %u Rx Seq %u -> %u", (unsigned)serialnumber, (unsigned)sequence_prev, (unsigned)sequence);
+
+    const bool bad_sequence = ((sequence_prev+1) != sequence);
+    if (bad_sequence) {
+        // we've dropped a packet somewhere
+        // TODO: handle lost packets
+        //gcs().send_text(MAV_SEVERITY_DEBUG, "PDCAN: %u Rx bad sequence ", (unsigned)serialnumber);
+    }
+
+    sequence_prev = sequence;
+
+    if (index >= device_count) {
+        // not sure how this can happen, but lets protect the array bounds just in case
+        return;
+    }
+
+    switch ((PacketType)frame.data[1]) {
+    case PacketType::STREAM_FAULT:
+        MPPT_devices[index].faults = (Fault_Flags)frame.data[2];
+        MPPT_devices[index].temperature = frame.data[3];
+        break;
+
+    case PacketType::STREAM_INPUT:
+        MPPT_devices[index].input.voltage = fixed2float(UINT16_VALUE(frame.data[2], frame.data[3]));
+        MPPT_devices[index].input.current = fixed2float(UINT16_VALUE(frame.data[4], frame.data[5]));
+        MPPT_devices[index].input.power = fixed2float(UINT16_VALUE(frame.data[6], frame.data[7]), 4); // NOTE: this is using 12:4 fixed point
+        break;
+    case PacketType::STREAM_OUTPUT:
+        MPPT_devices[index].output.voltage = fixed2float(UINT16_VALUE(frame.data[2], frame.data[3]));
+        MPPT_devices[index].output.current = fixed2float(UINT16_VALUE(frame.data[4], frame.data[5]));
+        MPPT_devices[index].output.power = fixed2float(UINT16_VALUE(frame.data[6], frame.data[7]), 4); // NOTE: this is using 12:4 fixed point
+        break;
+
+    case PacketType::FAULT: {
+        const Fault_Flags new_single_fault = (Fault_Flags)(~((uint8_t)MPPT_devices[index].faults & frame.data[2]));
```

Comment:
What makes this a single fault?

Reply:
I updated the logic to clarify and added a comment.